client/src/routes/[username]/[comic_id]/[chapter_id]/chapter-settings/+page.svelte

```diff
@@ -0,0 +1,273 @@
+    import { invalidate } from "$app/navigation";
+
+    export let data;
+
+    $: ({chapter, comic_id} = data);
+
+    const acceptedFileTypes = ["image/png", "image/jpeg", "image/jpg"];
+    
+    $: currentChapterNumber = chapter.number;
+
+    let pages: any[] = [];
+
+    function addImagesFromInput(e: Event) {
+        const input = e.target as HTMLInputElement;
+        for (const file of input.files || []) {
+            addPage(file)
+        }
+    }
+
+    function addPage(file: File) {
+        // TODO: limit file size
+        if (!acceptedFileTypes.includes(file.type)) return;
+        const reader = new FileReader();
+        reader.addEventListener("load", () => {
+            pages.push({
+              file,
+              path: reader.result,
+            });
+            pages = pages;
+        });
+        reader.readAsDataURL(file);
+    }
+
+    async function deleteServerChapterPage(page_id: string) {
+        console.log(page_id);
+        const deleteRes = await fetch(`http://localhost:6060/api/v1/comics/chapters/pages/${page_id}`, {
+          method: "DELETE",
+          credentials: "include",
+        });
+        console.log(deleteRes.statusText);
+        chapter.pages.splice(chapter.pages.findIndex((p) => p.id === page_id), 1);
+        chapter.pages = chapter.pages;
+        chapter = chapter;
+    }
+
+    async function deleteClientChapterPage(page: any) {
+        pages.splice(pages.findIndex((p) => p.path === page.path), 1);
+        pages = pages;
+    }
+    
+    async function updateChapter(e: SubmitEvent) {
+        const form = new FormData(e.target as HTMLFormElement);
+        let updateChapter: UpdateChapter = {
+          title: null,
+          description: null,
+          number: null,
+        };
+
+        const newPages = pages.reduce((result, page, i) => {
+          if (!page.file) return result;
+
+           result.push({
+            file: page.file,
+            number: chapter.pages.length + i + 1,
+            path: page.path,
+          });
+
+          return result;
+        }, []);
+
+        // TODO: add option to change desc and title
+        
+        if (currentChapterNumber !== chapter.number) {
+            updateChapter.number = currentChapterNumber;
+        }
+
+        for (const page of newPages) {
+          const createForm = new FormData();
+
+          createForm.append("number", page.number);
+
+          createForm.append("image", page.file);
+          
+          const createRes = await fetch(`http://localhost:6060/api/v1/comics/${comic_id}/chapters/${chapter.id}/pages`, {
+            method: "POST",
+            credentials: "include",
+            body: createForm,
+          });
+
+          // TODO: handle error
+          console.log(createRes.statusText);
+
+          invalidate("chapter-info");
+          pages.splice(pages.findIndex((p) => p.path === page.path) - 1, 1);
+          pages = pages;
+          chapter.pages = chapter.pages;
```

Comment:
this looks dubious, why are you not awaiting the promise from invalidate? why are you doing something *after* the invalidate?

Reply:
والله يا الامير مشارييي

Reply:
HAHAHAHAHAHAHAHA

Reply:
but for real, I don't know whether doing stuff after the invalidate is a bad thing

Reply:
my understanding is that since it reruns the `load` function, the props will have a new state, and changing the state after the invalidate is pointless since you will have new state by then, so your changes will be valid for like a few milliseconds. 

Not sure if my understanding is correct here

Reply:
Maybe remove the `chapter.pages = chapter.pages;` one, but the other one is ok I think

Reply:
Actually it does even go through the whole thing and removes all items, so it's actually useless

Reply:
not sure what you mean by this, but the code being useless makes sense and that's what I meant originally

Reply:
I would just await the invalidate at the end and not put anything after it, I need to experiment with it to understand, docs aren't so good

Reply:
I just tested it now, and it has a loading look when doing it like this:
```ts
          await invalidate("chapter-info");
          pages.splice(pages.findIndex((p) => p.path === page.path) - 1, 1);
          pages = pages;
```
but when it's like this:
```ts
          await invalidate("chapter-info");
```
it doesn't remove the client side pages, and just appends the server side ones

but doing it this way:
```ts
          pages.splice(pages.findIndex((p) => p.path === page.path) - 1, 1);
          pages = pages;
          await invalidate("chapter-info");
```
has a worse loading sequence, it removes the image then the server side one shows after, but the first one is instantaneous, so not jarring like this one

Reply:
welp, testing and benchmarking is above all.